egui_demo_lib/src/apps/demo/misc_demo_window.rs

```diff
@@ -339,28 +339,47 @@ enum Action {
 
 #[derive(Clone, Default)]
 #[cfg_attr(feature = "persistence", derive(serde::Deserialize, serde::Serialize))]
-struct Tree(Vec<Tree>);
+struct Tree(String, SubTree);
 
 impl Tree {
     pub fn demo() -> Self {
-        Self(vec![
-            Tree(vec![Tree::default(); 4]),
-            Tree(vec![Tree(vec![Tree::default(); 2]); 3]),
-        ])
+        Self(
+            String::default(),
+            SubTree(vec![
+                SubTree(vec![SubTree::default(); 4]),
+                SubTree(vec![SubTree(vec![SubTree::default(); 2]); 3]),
+            ]),
+        )
     }
     pub fn ui(&mut self, ui: &mut Ui) -> Action {
-        self.ui_impl(ui, 0, "root")
+        self.1.ui(ui, 0, "root", &mut self.0)
     }
+}
 
-    fn ui_impl(&mut self, ui: &mut Ui, depth: usize, name: &str) -> Action {
-        CollapsingHeader::new(name)
+#[derive(Clone, Default)]
+#[cfg_attr(feature = "persistence", derive(serde::Deserialize, serde::Serialize))]
+struct SubTree(Vec<SubTree>);
```

Comment:
Why do we need both `Tree` and `SubTree`?

Reply:
I preferred to split in order to avoid changing too muche the code and having the selected string everywhere, but we can change it if you prefer